### PR TITLE
Fix tests for 7.7.2

### PR DESCRIPTION
**Pull request by johnmbw**
I think some of the fixes we backported didn't pull over all the code needed.

This at least gets the tests compiling again on the 7.7.2 branch.

### CODE REVIEW

**Review by richardbadman — APPROVED** [review]

**Comment by johnmbw** [comment]
I ran this locally (it took 51 minutes!) and some of the tests failed.  Going to look into which ones failed, but might just be flakiness.

File: solr/core/src/java/org/apache/solr/search/facet/TopDocsAgg.java [hunk 1]
```diff
@@ -172,19 +174,9 @@ public SlotAcc createSlotAcc(FacetContext fcontext, int numDocs, int numSlots) t
...
-      String[] fl = null;
-      if (fieldsArg instanceof List) {
-        List lst = (List)fieldsArg;
-        fl = (String[])lst.toArray(new String[lst.size()]);
-      } else {
-        fl = (String[])fieldsArg;
-      }
-      returnFields = new SolrReturnFields(fl, fcontext.req);
+      returnFields = rb.rsp.getReturnFields();
+      returnFields = buildReturnFields(fcontext.req, rb.rsp);
```
Comment (timatbw) [inline solr/core/src/java/org/apache/solr/search/facet/TopDocsAgg.java:178]: not sure why this is setting the local var twice?

Comment (johnmbw) [reply solr/core/src/java/org/apache/solr/search/facet/TopDocsAgg.java:178]: huh - think I must have missed some code during a refactor.  I'll take a closer look and tweak it.

File: solr/test-framework/src/java/org/apache/solr/EmbeddedSolrServerTestBase.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+import com.google.common.io.ByteStreams;
+
+abstract public class EmbeddedSolrServerTestBase extends SolrTestCaseJ4 {
+
+  protected static final String DEFAULT_CORE_NAME = "collection1";
+
+  public static EmbeddedSolrServer client = null;
+
+  @After
+  public synchronized void afterClass() throws Exception {
```
Comment (timatbw) [inline solr/test-framework/src/java/org/apache/solr/EmbeddedSolrServerTestBase.java:54]: method name seems confusing as this is an instance method (and has `@After` too)

File: solr/test-framework/src/java/org/apache/solr/EmbeddedSolrServerTestBase.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+  }
+
+  /**
+   * Create a new solr client. Subclasses should override for other options.
+   */
+  public EmbeddedSolrServer createNewSolrClient() {
+    return new EmbeddedSolrServer(h.getCoreContainer(), DEFAULT_CORE_NAME) {
+      @Override
+      public void close() {
+        // do not close core container
```
Comment (timatbw) [inline solr/test-framework/src/java/org/apache/solr/EmbeddedSolrServerTestBase.java:78]: 😄  see my ticket https://issues.apache.org/jira/browse/SOLR-15085 although ofc that's only in master and 8.x

File: solr/test-framework/src/java/org/apache/solr/EmbeddedSolrServerTestBase.java (hunk 2 above)
Comment (johnmbw) [reply solr/test-framework/src/java/org/apache/solr/EmbeddedSolrServerTestBase.java:54]: This was just copied over from 8.6 (as we'd missed it from the original backport)

Comment (johnmbw) [reply solr/test-framework/src/java/org/apache/solr/EmbeddedSolrServerTestBase.java:54]: Though actually it was maybe from the original hash where we got the rest of the patch from.

File: solr/core/src/test/org/apache/solr/search/facet/TestJsonFacets.java [hunk 4]
```diff
@@ -3226,6 +3269,26 @@ private static String str(AVLTreeDigest digest) {
...
+
+  private static boolean doesFacetMethodSupportTermsPrefix() {
+    return FacetField.FacetMethod.DEFAULT_METHOD != FacetField.FacetMethod.DVSTRING;
+  }
+
+  private static boolean doesFacetMethodSupportSortedSetDocValues() {
+    return FacetField.FacetMethod.DEFAULT_METHOD != FacetField.FacetMethod.DVSTRING;
+  }
+
+  private static boolean doesFacetMethodSupportFieldsWithoutDocValues() {
```
Comment (timatbw) [inline solr/core/src/test/org/apache/solr/search/facet/TestJsonFacets.java]: I'm curious how these tests work for other facet methods like dvhash? And other methods that don't support all params like the above checks say

File: solr/core/src/java/org/apache/solr/search/facet/TopDocsAgg.java (hunk 1 above)
Comment (johnmbw) [reply solr/core/src/java/org/apache/solr/search/facet/TopDocsAgg.java:178]: I've sorted that out now (as a new commit I'll merge into the original later)

File: solr/core/src/test/org/apache/solr/search/facet/TestJsonFacets.java [hunk 5]
```diff
@@ -3226,6 +3269,26 @@ private static String str(AVLTreeDigest digest) {
...
+
+  private static boolean doesFacetMethodSupportNumericFields() {
+    return FacetField.FacetMethod.DEFAULT_METHOD != FacetField.FacetMethod.DVSTRING;
+  }
+
+  private static boolean doesFacetMethodSupportTermsPrefix() {
+    return FacetField.FacetMethod.DEFAULT_METHOD != FacetField.FacetMethod.DVSTRING;
+  }
+
+  private static boolean doesFacetMethodSupportSortedSetDocValues() {
```
Comment (timatbw) [inline solr/core/src/test/org/apache/solr/search/facet/TestJsonFacets.java]: Hmm, it should work.. dvstring supports multi-valued string fields at least (using SortedSetDocValues)

File: solr/test-framework/src/java/org/apache/solr/EmbeddedSolrServerTestBase.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+import org.apache.solr.common.util.ContentStream;
+import org.apache.solr.common.util.ContentStreamBase;
+import org.apache.solr.common.util.ContentStreamBase.ByteArrayStream;
+import org.apache.solr.util.ExternalPaths;
+import org.junit.After;
+import org.junit.AfterClass;
+
+import com.google.common.io.ByteStreams;
+
+abstract public class EmbeddedSolrServerTestBase extends SolrTestCaseJ4 {
```
Comment (timatbw) [inline solr/test-framework/src/java/org/apache/solr/EmbeddedSolrServerTestBase.java:47]: is this used? sorry if I missed it in the diff

**Review by timatbw — COMMENTED** [review]
It's a shame the TestJsonFacets needs so much change, as it could cause conflicts if we try to apply these patches to a newer upstream branch. But if 7.x is basically frozen now, maybe that's not an issue. It'll be the changes needed to 8.x that are perhaps more of a concern, if they diverge too much from upstream.

Maybe I should see if there's a way to either make dvstring support more of the functionality, or make that test have fewer assumptions about what works on all facet methods.

File: solr/test-framework/src/java/org/apache/solr/EmbeddedSolrServerTestBase.java (hunk 6 above)
Comment (johnmbw) [reply solr/test-framework/src/java/org/apache/solr/EmbeddedSolrServerTestBase.java:47]: Not in the code I wrote.  This whole class was missing from a patch we backported - the tests wouldn't compile without it.

File: solr/core/src/test/org/apache/solr/search/facet/TestJsonFacets.java (hunk 4 above)
Comment (johnmbw) [reply solr/core/src/test/org/apache/solr/search/facet/TestJsonFacets.java]: I _think_ they build an uninverted index from the terms in that case?

File: solr/core/src/test/org/apache/solr/search/facet/TestJsonFacets.java (hunk 5 above)
Comment (johnmbw) [reply solr/core/src/test/org/apache/solr/search/facet/TestJsonFacets.java]: It was failing on `DocValues.getSorted`, but maybe it's something about `sf.multiValued()` not being set for some reason.  I can have a closer look to see if I can figure it out.

**Comment by johnmbw** [comment]
mincount=0 is probably tricky, but I guess supporting prefix would be possible?

**Review by timatbw — APPROVED** [review]
OK lgtm.. I'm still surprised dvhash doesn't trip the tests since it doesn't support prefix or mincount 0 either, but /shrug/

**Comment by johnmbw** [comment]
Oh it looks like it only actually uses dvhash when it's not mincount=0 or has a prefix:
```
if (mincount > 0 && prefix == null && (ntype != null || method == FacetMethod.DVHASH)) {
        // TODO can we auto-pick for strings when term cardinality is much greater than DocSet cardinality?
        //   or if we don't know cardinality but DocSet size is very small
        return new FacetFieldProcessorByHashDV(fcontext, this, sf);
      } 
```
Guess we could do similar for this?

**Comment by johnmbw** [comment]
Ok I've changed that to behave the same way as it does for `dvhash` and just don't use it when it doesn't make sense.  Means the tests pass without any modification now.

**Review by timatbw — APPROVED** [review]
Great!

**Pull request merged**